### PR TITLE
add setters/getters in nn and autoencoders

### PR DESCRIPTION
**Pull request by sanuj**
Not merge ready. I'm getting [this error](http://pastebin.com/mDkMqaa6) with `PythonModular=ON`. It's an easy compiler error but I'm not able to debug it.

### CODE REVIEW

File: doc/ipython-notebooks/neuralnets/autoencoders.ipynb [hunk 1]
```diff
@@ -1,471 +1,484 @@
 {
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Deep Autoencoders"
+   ]
+  },
```
Comment (karlnapf) [inline doc/ipython-notebooks/neuralnets/autoencoders.ipynb]: something is very weird here. Why all these changes in the diff?
I expected single line changes...ideas?

File: src/shogun/io/NeuralNetworkFileReader.cpp [hunk 2]
```diff
@@ -138,36 +138,36 @@ CNeuralNetwork* CNeuralNetworkFileReader::parse_network(json_object* json_networ
 		{
 			const char* method = json_object_get_string(iter.val);
 			if (string_equal(method, "NNOM_LBFGS"))
-				network->optimization_method = NNOM_LBFGS;
```
Comment (karlnapf) [inline src/shogun/io/NeuralNetworkFileReader.cpp]: yes, like this is what the notebook diff should look like

File: src/shogun/io/NeuralNetworkFileReader.cpp [hunk 3]
```diff
@@ -138,36 +138,36 @@ CNeuralNetwork* CNeuralNetworkFileReader::parse_network(json_object* json_networ
...
+			network->set_gd_learning_rate(json_object_get_double(iter.val));
 		else if (string_equal(iter.key, "gd_learning_rate_decay"))
-			network->gd_learning_rate_decay = json_object_get_double(iter.val);
+			network->set_gd_learning_rate_decay(json_object_get_double(iter.val));
 		else if (string_equal(iter.key, "gd_momentum"))
-			network->gd_momentum = json_object_get_double(iter.val);
+			network->set_gd_momentum(json_object_get_double(iter.val));
 		else if (string_equal(iter.key, "gd_error_damping_coeff"))
-			network->gd_error_damping_coeff = json_object_get_double(iter.val);
+			network->set_gd_error_damping_coeff(json_object_get_double(iter.val));
```
Comment (karlnapf) [inline src/shogun/io/NeuralNetworkFileReader.cpp]: these all look good

File: src/shogun/neuralnets/Autoencoder.h [hunk 4]
```diff
@@ -165,6 +165,44 @@ class CAutoencoder : public CNeuralNetwork
...
+	 *
+	 * If set to AENT_GAUSSIAN, gaussian noise with zero mean and noise_parameter
+	 * standard deviation is added to the inputs.
+	 *
+	 * Default value is AENT_NONE
+	 * @param _noise_type noise type for denoising autoencoders
+	 */
+	void set_noise_type(EAENoiseType _noise_type)
+	{
+		this->noise_type = _noise_type;
```
Comment (karlnapf) [inline src/shogun/neuralnets/Autoencoder.h]: please dont use "this" pointer

Comment (karlnapf) [inline src/shogun/neuralnets/Autoencoder.h]: mmh.
Usually, members in shogun have an "m_" prefix, which avoid the need for underscores in setter variables....
But well, its ok for now
@lisitsyn

File: src/shogun/neuralnets/Autoencoder.h [hunk 5]
```diff
@@ -181,7 +219,18 @@ class CAutoencoder : public CNeuralNetwork
 	template<class T>
 	SGVector<T> get_section(SGVector<T> v, int32_t i);
 
-public:
+protected:
+	/** For contractive autoencoders [Rifai, 2011], a term:
```
Comment (karlnapf) [inline src/shogun/neuralnets/Autoencoder.h]: Did you write these comments? Or copied from somewhere?

File: src/shogun/neuralnets/NeuralNetwork.h [hunk 6]
```diff
@@ -231,6 +231,231 @@ friend class CDeepBeliefNetwork;
...
+	 * default value is 0.0
+	 * @param _l2_coefficient l2_coefficient
+	 */
+	void set_l2_coefficient(float64_t _l2_coefficient)
+	{
+		this->l2_coefficient = _l2_coefficient;
+	}
+
+	/** Returns L2 coefficient */
+	float64_t get_l2_coefficient()
```
Comment (karlnapf) [inline src/shogun/neuralnets/NeuralNetwork.h]: getters should be const

**Comment by sanuj** [comment]
@karlnapf can you review and merge?

**Comment by karlnapf** [comment]
Travis error unrelated, refactor patch, unit tests pass, diff looks good, merging

**Pull request merged**